src/hotspot/share/utilities/istream.hpp

```diff
@@ -0,0 +1,447 @@
+
+#include "memory/allocation.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/macros.hpp"
+#include "utilities/ostream.hpp"
+
+// Input streams for reading line-oriented textual data. These streams
+// treat newline '\n' very differently from all other bytes.  Carriage
+// return '\r' is just another bit of whitespace, although it is
+// removed just before newline.
+//
+// Null '\0' is just a data byte, although it also terminates C
+// strings; the `current_line` function adds a null after removing any
+// line terminator but does not specially process any nulls embedded
+// in the line.
+//
+// There are sizing access functions which allow lines to contain
+// null, but the simpler function assumes null termination, and thus
+// lines containing null will "look" shorter when viewed as C strings.
+// Use the sizing access functions if you care about this.
+//
+// Formatting guidelines:
+//
+// Configuration data should be line-oriented.  It should be readable
+// by humans (though perhaps with difficulty).  It should be easily
+// processed by text editors and by widely available text processing
+// tools such as grep, sed, and awk.
+//
+// Configuration data should not require "compilers" to generate, if
+// possible.  It should be editable by hand, if possible.  In cases
+// where binary data is strongly required, pick a binary format
+// already native to Hotspot, such as classfile, jar, or jmod.
+//
+// Each line should be separately parseable; the parsing can be ad
+// hoc.  For constructs inherently larger than single lines (such as
+// complex method configuration information), try to use a structuring
+// principle that allows "leaf" data to be line-oriented, and delimits
+// that data with markup lines of some sort.  Try to pick a
+// line-friendly version of a standard format like XML or Markdown.
+// JSON is somewhat problematic because there is no line-friendly leaf
+// syntax: everything at the leaves must be a quoted string in JSON.
+//
+// Use simple parsing via scanf-like formats for simple applications.
+// But, keep in mind that these formats may lose data when applied to
+// unusual strings, such as class names that contain spaces, or method
+// names that contain punctuation.  For more robust transmission of
+// potentially unusual names, consider wrapping them in XML-flavored
+// lines like <tag attr='pay load'/>.
+//
+// Note: Input streams are never MT-safe.
+
+class inputStream : public CHeapObjBase {
+ public:
+  class Input;
+
+ private:
+  NONCOPYABLE(inputStream);
+
+  static constexpr size_t SMALL_SIZE =  240 DEBUG_ONLY(*0 + 10);
+  static constexpr size_t BIG_SIZE   = 2048 DEBUG_ONLY(*0 + 20);
+
+ protected:
+  // Values for _input_state, to distinguish some phases of history:
+  // Do we need to read more input (NTR)?  Did we see EOF already?
+  // Was there an error getting input or allocating buffer space?
+  enum IState { NTR_STATE, EOF_STATE, ERR_STATE };
```

Comment:
Enum class is preferable.

Reply:
Is there an adopted style for the enumerators? I couldn't find it in the hotspot style guide. There's quite a variation today. E.g..

```
enum class vmSymbolID : int {  NO_SID = 0,   ....};
enum class DefaultsLookupMode { find, skip };
```

Reply:
According to my interpretation, the relevant style guide passage is about constants:

> * Constant names may be upper-case or mixed-case, according to historical necessity.  (Note: There are many examples of constants with lowercase names.)

Basically, “styles vary, follow precedent”.  I followed the precedent of `JavaThreadStatus` and others, as noted elsewhere in this PR.  OS-level constants have old-looking names like `EOF` and `FILENAME_MAX`.  That’s also the standard for Java constant names.